src/main/java/graphql/execution/AsynchronousExecutionStrategy.java

```diff
@@ -0,0 +1,223 @@
+            future = future.thenCombineAsync(resolveFieldFuture,(t,executionResult)-> t);
+        }
+
+        return new ExecutionResultImpl(future.thenApplyAsync(t -> results),executionContext.getErrors());
+    }
+
+    protected CompletionStage<ExecutionResult> resolveFieldAsync(ExecutionContext executionContext, ExecutionParameters parameters, List<Field> fields) {
+        GraphQLObjectType type = parameters.typeInfo().castType(GraphQLObjectType.class);
+        GraphQLFieldDefinition
+                        fieldDef = getFieldDef(executionContext.getGraphQLSchema(), type, fields.get(0));
+
+        Map<String, Object> argumentValues = valuesResolver.getArgumentValues(fieldDef.getArguments(), fields.get(0).getArguments(), executionContext.getVariables());
+
+        GraphQLOutputType fieldType = fieldDef.getType();
+        DataFetchingFieldSelectionSet fieldCollector = DataFetchingFieldSelectionSetImpl
+                        .newCollector(executionContext, fieldType, fields);
+
+        DataFetchingEnvironment environment = new DataFetchingEnvironmentImpl(
+                        parameters.source(),
+                        argumentValues,
+                        executionContext.getRoot(),
+                        fields,
+                        fieldType,
+                        type,
+                        executionContext.getGraphQLSchema(),
+                        executionContext.getFragmentsByName(),
+                        executionContext.getExecutionId(),
+                        fieldCollector);
+
+        Instrumentation instrumentation = executionContext.getInstrumentation();
+
+        InstrumentationContext<ExecutionResult>
+                        fieldCtx = instrumentation.beginField(new FieldParameters(executionContext, fieldDef, environment));
+
+        InstrumentationContext<Object> fetchCtx = instrumentation.beginFieldFetch(new FieldFetchParameters(executionContext, fieldDef, environment));
+        Object resolvedValue = null;
+
+        CompletableFuture<Object> dataFetcherResult = null;
+        try {
+            resolvedValue = fieldDef.getDataFetcher().get(environment);
+
+            if(resolvedValue instanceof CompletionStage) {
+                dataFetcherResult = (CompletableFuture) resolvedValue;
+            } else {
+                dataFetcherResult = CompletableFuture.completedFuture(resolvedValue);
+            }
+        } catch (Exception e) {
+            log.warn("Exception while fetching data", e);
+            dataFetcherResult = new CompletableFuture();
+            dataFetcherResult.completeExceptionally(e);
+        }
+
+        return dataFetcherResult.handleAsync((value,th)-> {
+            if(th != null) {
+                log.warn("Exception while fetching data", th);
+                handleDataFetchingException(executionContext, fieldDef, argumentValues, new ExecutionException(th));
+                fetchCtx.onEnd(th);
+            }
+
```

Comment:
The Instrumentation is not called back here and hence is lost

            fetchCtx.onEnd(resolvedValue);

- [ ] Add in intrumentation call